compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,505 @@
+use crate::assert::expr_if_not;
+use rustc_ast::{
+    ptr::P,
+    token::{self, Delimiter},
+    tokenstream::{DelimSpan, TokenStream, TokenTree},
+    Block, BorrowKind, Expr, ExprKind, ItemKind, LocalKind, MacArgs, MacCall, MacDelimiter,
+    Mutability, Path, PathSegment, Stmt, StmtKind, StructRest, UseTree, UseTreeKind, DUMMY_NODE_ID,
+};
+use rustc_ast_pretty::pprust;
+use rustc_expand::base::ExtCtxt;
+use rustc_span::{
+    symbol::{sym, Ident, Symbol},
+    Span, DUMMY_SP,
+};
+
+pub(super) struct Context<'cx, 'a> {
+    span: Span,
+    captures: Vec<Capture>,
+    cx: &'cx ExtCtxt<'a>,
+}
+
+impl<'cx, 'a> Context<'cx, 'a> {
+    pub(super) fn new(cx: &'cx ExtCtxt<'a>, span: Span) -> Self {
+        Self { cx, captures: Vec::new(), span }
+    }
+
+    /// Builds the whole `assert!` expression.
+    ///
+    /// {
+    ///    use ::core::asserting::{ ... };
+    ///
+    ///    let mut __capture0 = Capture::Unevaluated;
+    ///    ...
+    ///    ...
+    ///    ...
+    ///
+    ///    if !{
+    ///       ...
+    ///       ...
+    ///       ...
+    ///    } {
+    ///        panic!(
+    ///            "Assertion failed: ... \n With expansion: ...",
+    ///            (&__capture0).printer_kind().new(" ... ", __capture0),
+    ///            ...
+    ///            ...
+    ///            ...
+    ///        );
+    ///    }
+    /// }
+    pub(super) fn build(mut self, mut cond_expr: P<Expr>, panic_path: Path) -> P<Expr> {
+        let expr_str = escape_to_fmt(&pprust::expr_to_string(&cond_expr));
+        let mut fmt_str = String::new();
+        self.manage_cond_expr(&mut cond_expr, &mut fmt_str);
+        let initial_imports = self.build_initial_imports();
+        let panic = self.build_panic(&expr_str, &fmt_str, panic_path);
+
+        let Self { captures, cx, .. } = self;
+
+        let mut stmts = vec![initial_imports];
+        stmts.extend(captures.into_iter().map(|c| c.decl));
+        stmts.push(cx.stmt_expr(expr_if_not(cx, DUMMY_SP, cond_expr, panic, None)));
+        cx.expr_block(cx.block(DUMMY_SP, stmts))
+    }
+
+    /// Initial **trait** imports
+    ///
+    /// use ::core::asserting::{ ... };
+    fn build_initial_imports(&self) -> Stmt {
+        self.cx.stmt_item(
+            DUMMY_SP,
```

Comment:
Spans for generated code will need to be set correctly before merging.
Both the location component and the hygiene component, see other built-in macros for examples.

Reply:
All span-related fields are now using `self.span` instead of `DUMMY_SP`. Is it OK?

```bash
$ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Zunpretty=expanded,hygiene src/test/ui/rfc-2011-nicer-assert-messages/feature-gate-generic_assert.rs
```

```
fn test /* 1344#0 */() {
    let copy_debug /* 1519#0 */ = CopyDebug /* 1516#0 */(1);
    {
        use ::core
        /*
        2#8
        */::asserting
        /*
        300#0
        */::{TryCaptureGeneric /* 204#0 */, TryCapturePrintable /* 205#0 */};
        let mut __capture0 /* 1736#7 */ =
            ::core /* 2#8 */::asserting /* 300#0 */::Capture /* 97#0 */::new
                    /* 891#0 */();
        if ::core /* 2#8 */::intrinsics /* 741#0 */::unlikely /* 1413#0
                        */(!({
                                let __local_bind /* 224#7 */ = copy_debug /* 1519#0 */;
                                (&::core /* 2#8 */::asserting /* 300#0 */::Wrapper /* 216#0
                                                    */(&__local_bind /* 224#7 */)).try_capture /* 1372#7
                                    */(&mut __capture0 /* 1736#7 */);
                                __local_bind /* 224#7 */
                            } == CopyDebug /* 1516#0 */(3))) {

                {
                    ::std /* 2#14 */::rt /* 1095#14 */::panic_fmt /* 957#14
                            */(::core /* 2#15 */::fmt /* 634#0 */::Arguments /* 71#0
                                */::new_v1 /* 1785#0
                                */(&["Assertion failed: copy_debug == CopyDebug(3)\nWith captures:\n  copy_debug = ",
                                        "\n"],
                            &[::core /* 2#15 */::fmt /* 634#0 */::ArgumentV1 /* 70#0
                                                */::new_debug /* 1784#0 */(&__capture0 /* 1736#7 */)]))
                }
            }
    };
}

/*
Expansions:
crate0::{{expn0}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Root
crate0::{{expn1}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: AstPass(StdImports)
crate0::{{expn2}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate0::{{expn3}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "assert")
crate0::{{expn4}}: parent: crate0::{{expn2}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Derive, "Clone")
crate0::{{expn5}}: parent: crate0::{{expn2}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Derive, "Copy")
crate0::{{expn6}}: parent: crate0::{{expn2}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Derive, "PartialEq")
crate0::{{expn7}}: parent: crate0::{{expn3}}, call_site_ctxt: #7, def_site_ctxt: #0, kind: Macro(Bang, "panic")
crate0::{{expn8}}: parent: crate0::{{expn7}}, call_site_ctxt: #10, def_site_ctxt: #0, kind: Macro(Bang, "$crate::panic::panic_2015")
crate0::{{expn9}}: parent: crate0::{{expn8}}, call_site_ctxt: #14, def_site_ctxt: #0, kind: Macro(Bang, "$crate::const_format_args")
crate0::{{expn10}}: parent: crate0::{{expn0}}, call_site_ctxt: #7, def_site_ctxt: #0, kind: Desugaring(CondTemporary)
crate1::{{expn1}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: AstPass(StdImports)
crate1::{{expn2}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "include_str")
crate1::{{expn211}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate1::{{expn440}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "include")
crate1::{{expn441}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "include")
crate2::{{expn223}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn618}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn642}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn652}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn657}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "atomic_int")
crate2::{{expn658}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "atomic_int")
crate2::{{expn661}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "atomic_int_ptr_sized")
crate2::{{expn701}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn717}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Attr, "derive")
crate2::{{expn2727}}: parent: crate0::{{expn0}}, call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "include")
crate2::{{expn17615}}: parent: crate2::{{expn661}}, call_site_ctxt: #13, def_site_ctxt: #0, kind: Macro(Bang, "atomic_int")
crate2::{{expn17616}}: parent: crate2::{{expn661}}, call_site_ctxt: #13, def_site_ctxt: #0, kind: Macro(Bang, "atomic_int")

SyntaxContexts:
#0: parent: #0, outer_mark: (crate0::{{expn0}}, Opaque)
#1: parent: #0, outer_mark: (crate0::{{expn1}}, Opaque)
#2: parent: #0, outer_mark: (crate0::{{expn1}}, Transparent)
#3: parent: #0, outer_mark: (crate1::{{expn1}}, Opaque)
#4: parent: #0, outer_mark: (crate0::{{expn4}}, Opaque)
#5: parent: #0, outer_mark: (crate0::{{expn5}}, Opaque)
#6: parent: #0, outer_mark: (crate0::{{expn6}}, Opaque)
#7: parent: #0, outer_mark: (crate0::{{expn3}}, Transparent)
#8: parent: #0, outer_mark: (crate0::{{expn3}}, Opaque)
#9: parent: #0, outer_mark: (crate1::{{expn2}}, Opaque)
#10: parent: #0, outer_mark: (crate0::{{expn7}}, Transparent)
#11: parent: #0, outer_mark: (crate0::{{expn7}}, Opaque)
#12: parent: #0, outer_mark: (crate2::{{expn17616}}, SemiTransparent)
#13: parent: #0, outer_mark: (crate2::{{expn661}}, SemiTransparent)
#14: parent: #0, outer_mark: (crate0::{{expn8}}, SemiTransparent)
#15: parent: #0, outer_mark: (crate0::{{expn9}}, Opaque)
#16: parent: #0, outer_mark: (crate2::{{expn657}}, SemiTransparent)
#17: parent: #0, outer_mark: (crate2::{{expn17615}}, SemiTransparent)
#18: parent: #0, outer_mark: (crate2::{{expn658}}, SemiTransparent)
#19: parent: #0, outer_mark: (crate0::{{expn10}}, Transparent)
*/
```